### PR TITLE
Cache optimization + refactoring

### PR DESCRIPTION
**Pull request by polonzer**
https://lykkex.atlassian.net/browse/LWDEV-9224

### CODE REVIEW

File: src/Lykke.Service.Assets/Controllers/v2/AssetGroupsController.cs [hunk 1]
```diff
@@ -1,30 +1,23 @@
...
 using Swashbuckle.AspNetCore.SwaggerGen;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
 
 namespace Lykke.Service.Assets.Controllers.V2
 {
     [Route("/api/v2/asset-groups")]
     public class AssetGroupsController : Controller
```
Comment (njannink) [inline src/Lykke.Service.Assets/Controllers/v2/AssetGroupsController.cs]: Apply [ApiController] attribute for default rest api controller behavior

Comment (njannink) [reply src/Lykke.Service.Assets/Controllers/v2/AssetGroupsController.cs]: Applies to all controllers

File: src/Lykke.Service.Assets/Modules/CacheModule.cs [hunk 2]
```diff
@@ -58,21 +46,24 @@ protected override void Load(ContainerBuilder builder)
...
 
         private void RegisterRedis(ContainerBuilder builder)
         {
-            System.Threading.ThreadPool.SetMinThreads(100, 100);
-            var options = ConfigurationOptions.Parse(_settings.CurrentValue.AssetsService.RadisSettings.RedisConfiguration);
+            System.Threading.ThreadPool.SetMinThreads(300, 300);
+            var options = ConfigurationOptions.Parse(_settings.RedisSettings.Configuration);
             options.ReconnectRetryPolicy = new ExponentialRetry(3000, 15000);
 
             var redis = ConnectionMultiplexer.Connect(options);
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.Assets/Modules/CacheModule.cs]: set PreserveAsyncOrder = false;
https://stackoverflow.com/a/30912888/8592182

Comment (polonzer) [reply src/Lykke.Service.Assets/Modules/CacheModule.cs]: This option is deprecated.

File: src/Lykke.Service.Assets/Cache/AssetsForClientCacheManager.cs [hunk 3]
```diff
@@ -1,47 +1,46 @@
...
             _redisDatabase = redisDatabase;
+            _expiration = expiration;
             _log = logFactory.CreateLog(this);
+            _partitionKey = partitionKey;
         }
 
         public async Task ClearCacheAsync(string reason)
         {
-            RedisKey[] keys = _redisServer.Keys(pattern: $"{_settings.InstanceName}{PatternClient}*", pageSize: 1000).ToArray();
+            RedisKey[] keys = _redisServer.Keys(pattern: $"{_partitionKey}*", pageSize: 1000).ToArray();
```
[the remark this answers is not in the document]
Comment (polonzer) [reply src/Lykke.Service.Assets/Cache/AssetsForClientCacheManager.cs]: This is totally not what I want to see in the source code. But it is already there for a while. I'll refactor it in the next task. Moreover, I'll delete this class at all. But next time :)

File: src/Lykke.Service.Assets/Cache/CacheSerializer.cs [hunk 4]
```diff
@@ -1,15 +1,25 @@
-﻿namespace Lykke.Service.Assets.Cache
+﻿using System.IO;
+
+namespace Lykke.Service.Assets.Cache
 {
     internal static class CacheSerializer
```
[the remark this answers is not in the document]
Comment (polonzer) [reply src/Lykke.Service.Assets/Cache/CacheSerializer.cs]: I don't think it is suitable here.

File: src/Lykke.Service.Assets/Modules/CqrsModule.cs [hunk 5]
```diff
@@ -64,44 +59,39 @@ protected override void Load(ContainerBuilder builder)
...
                     ctx.Resolve<IDependencyResolver>(),
-#if DEBUG
-                    new MessagingEngine(ctx.Resolve<ILogFactory>(),
-                        new TransportResolver(new Dictionary<string, TransportInfo>
-                        {
-                            {"RabbitMq", new TransportInfo(rabbitMqSettings.Endpoint + virtualHost, rabbitMqSettings.UserName, rabbitMqSettings.Password, "None", "RabbitMq")}
-                        }),
-                        new RabbitMqTransportFactory(ctx.Resolve<ILogFactory>())),
-#else
                     new MessagingEngine(ctx.Resolve<ILogFactory>(),
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.Assets/Modules/CqrsModule.cs]: It is important to dispose MessagingEngine but in the current implementation is will never be disposed. Register it in the container.

File: src/Lykke.Service.Assets/Cache/CachedAssetCategoryService.cs [hunk 6]
```diff
@@ -2,39 +2,41 @@
...
-            return await _cache.GetAsync(id, () => _assetCategoryService.GetAsync(id));
+            return await _cache.GetAsync(id, 
+                async () => AutoMapper.Mapper.Map<AssetCategory>(await _assetCategoryService.GetAsync(id)));
         }
 
-        public async Task<IEnumerable<IAssetCategory>> GetAllAsync()
+        public async Task<IEnumerable<AssetCategory>> GetAllAsync()
         {
-            return await _cache.GetListAsync(AllEntitiesKey, () => _assetCategoryService.GetAllAsync());
+            return await _cache.GetListAsync(AllEntitiesKey,
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.Assets/Cache/CachedAssetCategoryService.cs]: You don't need await here

File: src/Lykke.Service.Assets/Cache/CachedAssetCategoryService.cs [hunk 7]
```diff
@@ -2,39 +2,41 @@
...
 
-            return await _assetCategoryService.AddAsync(assetCategory);
+            return AutoMapper.Mapper.Map<AssetCategory>(await _assetCategoryService.AddAsync(assetCategory));
         }
 
-        public async Task<IAssetCategory> GetAsync(string id)
+        public async Task<AssetCategory> GetAsync(string id)
         {
-            return await _cache.GetAsync(id, () => _assetCategoryService.GetAsync(id));
+            return await _cache.GetAsync(id,
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.Assets/Cache/CachedAssetCategoryService.cs]: You don't need await here

File: src/Lykke.Service.Assets/Settings/RedisSettings.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Lykke.Service.Assets.Settings
+{
+    public partial class ApplicationSettings
+    {
+        public class RedisSettings
+        {
+            public string Configuration { get; set; }
+            public string Instance { get; set; }
+            public TimeSpan Expiration { get; set; }
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.Assets/Settings/RedisSettings.cs]: I would make the field optional and set some reasonable timeout.

**Pull request merged**